src/riak_core_gossip.erl

```diff
@@ -367,73 +371,287 @@ remove_from_cluster(Ring, ExitingNode) ->
     remove_from_cluster(Ring, ExitingNode, rand:seed(exrop, os:timestamp())).
 
 remove_from_cluster(Ring, ExitingNode, Seed) ->
-    % Get a list of indices owned by the ExitingNode...
-    AllOwners = riak_core_ring:all_owners(Ring),
-
     % Transfer indexes to other nodes...
     ExitRing =
-        case attempt_simple_transfer(Seed, Ring, AllOwners, ExitingNode) of
+        case attempt_simple_transfer(Seed, Ring, ExitingNode) of
             {ok, NR} ->
                 NR;
-            target_n_fail ->
+            _ ->
                 %% re-diagonalize
                 %% first hand off all claims to *any* one else,
                 %% just so rebalance doesn't include exiting node
+                Owners = riak_core_ring:all_owners(Ring),
                 Members = riak_core_ring:claiming_members(Ring),
-                Other = hd(lists:delete(ExitingNode, Members)),
-                TempRing = lists:foldl(
-                             fun({I,N}, R) when N == ExitingNode ->
-                                     riak_core_ring:transfer_node(I, Other, R);
-                                (_, R) -> R
-                             end,
-                             Ring,
-                             AllOwners),
-                riak_core_claim:claim_rebalance_n(TempRing, Other)
+                HN = hd(lists:delete(ExitingNode, Members)),
```

Comment:
I see that the code is not new, but are we absolutely sure that Members always has at least 2 elements, such that we have an element left to take as head?

Reply:
It is quite a long chain of functions - but we get here because at some stage there has been a call to riak_core-claimant:maybe_update_ring/5 - https://github.com/basho/riak_core/blob/1f7e204548148f7dda9727113838d40b559310e3/src/riak_core_claimant.erl#L1235-L1254

This checks that claiming_members is non-empty